### PR TITLE
Running seed will convert downloaded songs to opus .ogg format

### PR DESCRIPTION
**Pull request by taahamahdi**

### CODE REVIEW

File: src/scripts/download-new-songs.ts [hunk 1]
```diff
@@ -133,17 +133,88 @@ const downloadNewSongs = async (limit?: number) => {
...
+    ffmpegOpusJob(mp3Files);
+    logger.info("Completed jobs of length", mp3Files.length);
+}
+
+const ffmpegOpusJob = (files: Array<string>) => {
+    files.forEach(async (mp3File) => {
+        try {
+            let oggFileWithPath: string = path.join(process.env.SONG_DOWNLOAD_DIR, `${path.basename(mp3File, ".mp3")}.ogg`);
+            if (fs.existsSync(oggFileWithPath)) {
+                return;
```
Comment (Brainicism) [inline src/scripts/download-new-songs.ts:153]: Should this be `continue;`?

File: src/scripts/download-new-songs.ts [hunk 2]
```diff
@@ -133,17 +133,88 @@ const downloadNewSongs = async (limit?: number) => {
...
+    logger.info(`Converting ${mp3Files.length} from mp3 to opus (in ogg container)`)
+
+    ffmpegOpusJob(mp3Files);
+    logger.info("Completed jobs of length", mp3Files.length);
+}
+
+const ffmpegOpusJob = (files: Array<string>) => {
+    files.forEach(async (mp3File) => {
+        try {
+            let oggFileWithPath: string = path.join(process.env.SONG_DOWNLOAD_DIR, `${path.basename(mp3File, ".mp3")}.ogg`);
```
Comment (Brainicism) [inline src/scripts/download-new-songs.ts:151]: Only explicitly annotate the type if it can't be inferred

File: src/scripts/download-new-songs.ts [hunk 3]
```diff
@@ -133,17 +133,88 @@ const downloadNewSongs = async (limit?: number) => {
...
+            if (fs.existsSync(oggFileWithPath)) {
+                return;
+            }
+
+            let oggPartWithPath = `${oggFileWithPath}.part`;
+            let oggFfmpegOutputStream: fs.WriteStream = fs.createWriteStream(oggPartWithPath);
```
Comment (Brainicism) [inline src/scripts/download-new-songs.ts:157]: Only explicitly annotate the type if it can't be inferred

File: src/scripts/download-new-songs.ts [hunk 4]
```diff
@@ -133,17 +133,88 @@ const downloadNewSongs = async (limit?: number) => {
...
+        }
+    })
+}
+
+const downloadAndConvertSongs = async (limit?: number) => {
+    if (!fs.existsSync(process.env.SONG_DOWNLOAD_DIR)) {
+        return logger.error("Song cache directory doesn't exist.");
+    }
+
+    let files: Array<string>;
```
Comment (Brainicism) [inline src/scripts/download-new-songs.ts:197]: Only explicitly annotate the type if it can't be inferred

File: src/scripts/download-new-songs.ts [hunk 5]
```diff
@@ -133,17 +133,88 @@ const downloadNewSongs = async (limit?: number) => {
...
+    const endingWithMp3Regex = new RegExp("\\.mp3$");
+    const mp3Files = files.filter((file) => file.match(endingWithMp3Regex));
+    logger.info(`Converting ${mp3Files.length} from mp3 to opus (in ogg container)`)
+
+    ffmpegOpusJob(mp3Files);
+    logger.info("Completed jobs of length", mp3Files.length);
+}
+
+const ffmpegOpusJob = (files: Array<string>) => {
+    files.forEach(async (mp3File) => {
```
Comment (Brainicism) [inline src/scripts/download-new-songs.ts:149]: This callback should be it's own async function that returns a Promise that we `await` on. Right now it looks like nothing is being waited on

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: src/scripts/download-new-songs.ts (hunk 1 above)
Comment (taahamahdi) [reply src/scripts/download-new-songs.ts:153]: I just checked, the equivalent for `continue` in a `forEach` is `return`

File: src/scripts/download-new-songs.ts [hunk 6]
```diff
@@ -61,10 +62,18 @@ const downloadSong = (id: string) => {
...
                 .pipe(cacheStream);
         } catch (e) {
-            await db.kmq("dead_links")
-                .insert({ vlink: id, reason: `Failed to retrieve video metadata. error = ${e}` });
-            reject(`Failed to retrieve video metadata. error = ${e}`);
-            return;
+            if (e !== "Video unavailable" && retries < 3) {
+                logger.error(`Retrying ${id} again...`)
+                downloadSong(id, ++retries);
+                resolve();
```
Comment (Brainicism) [inline src/scripts/download-new-songs.ts:68]: No guarantee `downloadSong` completed successfully, but you still `resolve`

File: src/scripts/download-new-songs.ts [hunk 7]
```diff
@@ -133,17 +146,88 @@ const downloadNewSongs = async (limit?: number) => {
...
     })
     await db.destroy();
-    logger.info(`Total songs downloaded: ${downloadCount}`);
+    logger.info(`Total songs downloaded: ${downloadCount}, (${deadLinksSkipped} dead links skipped)`);
+}
+
+const convertToOpus = async () => {
+    let files: Array<string>;
+    try {
+        files = await fs.promises.readdir(process.env.SONG_DOWNLOAD_DIR);
```
Comment (Brainicism) [inline src/scripts/download-new-songs.ts:155]: Let's just assume the directory exists, get rid of the try catch

File: src/scripts/download-new-songs.ts [hunk 8]
```diff
@@ -133,17 +146,88 @@ const downloadNewSongs = async (limit?: number) => {
...
+    }
+    catch (err) {
+        return logger.error(err);
+    }
+
+    const endingWithMp3Regex = new RegExp("\\.mp3$");
+    const mp3Files = files.filter((file) => file.match(endingWithMp3Regex));
+    logger.info(`Converting ${mp3Files.length} from mp3 to opus (in ogg container)`)
+
+    let opusJobs = [];
```
Comment (Brainicism) [inline src/scripts/download-new-songs.ts:165]: `let opusJobs = mp3Files.map((file) => ffmpegOpusJob(file))`

File: src/scripts/download-new-songs.ts [hunk 9]
```diff
@@ -133,17 +146,88 @@ const downloadNewSongs = async (limit?: number) => {
...
+    mp3Files.forEach((mp3File) => {
+        opusJobs.push(ffmpegOpusJob(mp3File));
+    });
+    Promise.all(opusJobs).then(() => {
```
Comment (Brainicism) [inline src/scripts/download-new-songs.ts:169]: await Promise.all

File: src/scripts/download-new-songs.ts [hunk 10]
```diff
@@ -133,17 +146,88 @@ const downloadNewSongs = async (limit?: number) => {
...
+
+    let oggPartWithPath = `${oggFileWithPath}.part`;
+    let oggFfmpegOutputStream = fs.createWriteStream(oggPartWithPath);
+
+    return new Promise((resolve, reject) => {
+        ffmpeg(`${process.env.SONG_DOWNLOAD_DIR}/${mp3File}`)
+            .format("opus")
+            .audioCodec("libopus")
+            .output(oggFfmpegOutputStream)
+            .on('end', () => {
```
Comment (Brainicism) [inline src/scripts/download-new-songs.ts:188]: use double quotes

File: src/scripts/download-new-songs.ts [hunk 11]
```diff
@@ -133,17 +146,88 @@ const downloadNewSongs = async (limit?: number) => {
...
+                    if (!fs.existsSync(oggFileWithPath)) {
+                        logger.error(`File ${oggFileWithPath} wasn't created. Ignoring...`);
+                        reject();
+                        return;
+                    }
+                    logger.info(`File ${oggFileWithPath} might have duplicate entries in db.`)
+                    reject();
+                }
+            })
+            .on('error', (transcodingErr) => {
```
Comment (Brainicism) [inline src/scripts/download-new-songs.ts:206]: use double quotes

**Review by Brainicism — CHANGES_REQUESTED** [review]

**Review by Brainicism — APPROVED** [review]

**Pull request merged**